src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -0,0 +1,249 @@
+            return null;
+        }
+
+        if (!isAllowedRequest(request)) {
+            return null;
+        }
+
+        try {
+            final Claims claims = jwtParser.parseClaimsJws(jwtToken).getBody();
+
+            final String subject = claims.getSubject();
+            if (subject == null) {
+                log.error("Valid jwt on behalf of token with no subject");
+                return null;
+            }
+
+            final String audience = claims.getAudience();
+            if (audience == null) {
+                log.error("Valid jwt on behalf of token with no audience");
+                return null;
+            }
+
+            final String issuer = claims.getIssuer();
+            if (!issuer.equals(clusterName)) {
+                log.error("The issuer of this OBO does not match the current cluster identifier");
+                return null;
+            }
+
+            List<String> roles = extractSecurityRolesFromClaims(claims);
+            String[] backendRoles = extractBackendRolesFromClaims(claims);
+
+            final AuthCredentials ac = new AuthCredentials(subject, roles, backendRoles).markComplete();
+
+            for (Entry<String, Object> claim : claims.entrySet()) {
+                ac.addAttribute("attr.jwt." + claim.getKey(), String.valueOf(claim.getValue()));
+            }
+
+            return ac;
+
+        } catch (WeakKeyException e) {
+            log.error("Cannot authenticate user with JWT because of ", e);
+        } catch (Exception e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Invalid or expired JWT token.", e);
+            }
+        }
+
+        return null;
+    }
+
+    private String extractJwtFromHeader(RestRequest request) {
+        String jwtToken = request.header(HttpHeaders.AUTHORIZATION);
+
+        if (jwtToken == null || jwtToken.isEmpty()) {
+            logDebug("No JWT token found in '{}' header", HttpHeaders.AUTHORIZATION);
+            return null;
+        }
+
+        if (!BEARER.matcher(jwtToken).matches()) {
+            return null;
+        }
+
+        if (jwtToken.toLowerCase().contains(BEARER_PREFIX)) {
+            jwtToken = jwtToken.substring(jwtToken.toLowerCase().indexOf(BEARER_PREFIX) + BEARER_PREFIX.length());
+        } else {
+            logDebug("No Bearer scheme found in header");
+            return null;
+        }
+
+        return jwtToken;
+    }
+
+    private void logDebug(String message, Object... args) {
+        if (log.isDebugEnabled()) {
+            log.debug(message, args);
+        }
+    }
+
+    public Boolean isAllowedRequest(final RestRequest request) {
+        Matcher matcher = PATTERN_PATH_PREFIX.matcher(request.path());
+        final String suffix = matcher.matches() ? matcher.group(2) : null;
+        if (request.method() == RestRequest.Method.POST && ON_BEHALF_OF_SUFFIX.equals(suffix)
+            || request.method() == RestRequest.Method.PUT && ACCOUNT_SUFFIX.equals(suffix)) {
```

Comment:
Why are POST/PUT methods not allowed with OBO tokens?  I'd recommend removing and creating a follow up issue to add this back / document how this works.

Reply:
It is only restricting the POST for OBO endpoint + PUT for account endpoint for limiting the usage of OBO for password changing. I believe issue https://github.com/opensearch-project/security/issues/2891 captured that.

Reply:
I've reactivated that issue as there was no decision or finding on the issue - there should be some kind of resolution and indication of the path forward.

If we are going to block how these APIs are called, break how they are detected break them out into individual functions `isCreateOnBehalfOfTokenRequest(...) {...}` `isUpdatePasswordRequest(...) {...}` and use public fields from those classes declared in this codebase.  Note; IMO this can be followed up seperately if you want to create an issue to track.

Reply:
I think we were in the alignment of closing that one during one of the standup is that since it meets the goal of blocking the access of these 2 endpoints. But yeah, I'm fine with either way, and if we re-open that up, we can follow-up on the original issue then.

Reply:
I just re-edit the description of issue #2891  to track the follow up work of refactoring